doppler/resource_config.go

```diff
@@ -84,7 +84,7 @@ func resourceConfigRead(ctx context.Context, d *schema.ResourceData, m interface
 
 	config, err := client.GetConfig(ctx, project, name)
 	if err != nil {
-		return diag.FromErr(err)
+		return handleNotFoundError(err, d)
```

Comment:
I don't think we want to mark resources as deleted if _any_ error occurs, just Not Found errors. Unfortunately, it doesn't like all of our our API endpoints are consistent about the returned HTTP code when a resource is not found (a 404 would be expected):

```
GET https://api.doppler.com/v3/projects/project?project=asdf

HTTP/1.1 400 Bad Request
date: Wed, 05 Apr 2023 12:55:01 GMT
...

#+RESPONSE
{
  "messages": [
    "Could not find requested project 'asdf'"
  ],
  "success": false
}
#+END
```

Before we can check these errors, I think we'll need to tweak the backend to return 404s for missing projects, environments, and configs: https://github.com/DopplerHQ/server/commit/378766906d9176c9effe92f5f0ee28facd89df97

Integrations and syncs already return 404s, which is good. Secrets are special so I left a separate comment for that. Your impl for service tokens looks perfect 👍

Reply:
Interesting. This does get picked up as a failure for projects (and we're already filtering to only mark them as deleted specifically for errors that have messages containing "Could not find requested"), but it seems to still fail in such a way that things don't work as expected:

```
doppler_project.rocket: Refreshing state... [id=rocket]
doppler_environment.rocket_dev: Refreshing state... [id=rocket.dev]
doppler_environment.rocket_prod: Refreshing state... [id=rocket.prod]
doppler_environment.rocket_stage: Refreshing state... [id=rocket.stage]
doppler_environment.rocket_ci: Refreshing state... [id=rocket.ci]
doppler_secret.elasticsearch: Refreshing state... [id=rocket.dev.ENDPOINT]
doppler_secret.test: Refreshing state... [id=rocket.dev.TEST]
doppler_config.rocket_ci_circleci: Refreshing state... [id=rocket.ci.ci_circleci]
doppler_config.rocket_ci_github: Refreshing state... [id=rocket.ci.ci_github]
doppler_service_token.rocket_ci_token: Refreshing state... [id=rocket.ci.20089d98-8ed6-41c2-b3ea-9ba36a3f68b1]
╷
│ Warning: Doppler Error: Could not find requested project 'rocket'
│
│   with doppler_project.rocket,
│   on main.tf line 22, in resource "doppler_project" "rocket":
│   22: resource "doppler_project" "rocket" {
│
│ Resource was not found, so it was removed from state and is being recreated.
│
│ (and 8 more similar warnings elsewhere)
╵
╷
│ Error: Doppler Error: Could not find requested project 'rocket'
│
│   with doppler_service_token.rocket_ci_token,
│   on main.tf line 79, in resource "doppler_service_token" "rocket_ci_token":
│   79: resource "doppler_service_token" "rocket_ci_token" {
│
╵
```

Reply:
Ah, I didn't notice the error message check in `handleNotFoundError` initially. I don't think we can rely on that message staying the same in the API, the 404 status code would be a much more reliable indicator.

Good find on the service token failure! With your latest updates, does Terraform appropriately recreate the service token on the new project in the scenario that you posted?

Reply:
Even though the API functions return `error`, that's just an interface that `APIError` conforms to. You should be able to use this patch to check if the returned error is an `APIError` and extract the status code: https://github.com/DopplerHQ/terraform-provider-doppler/pull/43#discussion_r1158836483

Reply:
Yep, that's already done I think (that comment was from yesterday and just got stuck as pending). I think the only thing that still needs discussing is https://github.com/DopplerHQ/terraform-provider-doppler/pull/43#discussion_r1158917091.